File: src/tasks/model/impl/CyclicalDependencyException.java
package tasks.model.impl;

import omnia.data.structure.List;
import omnia.data.structure.immutable.ImmutableList;

public class CyclicalDependencyException extends IllegalStateException {
  private final List<Object> cycle;

  CyclicalDependencyException(String msg, List<?> cycle) {
    super(buildMessage(msg, ImmutableList.copyOf(cycle)));
    this.cycle = ImmutableList.copyOf(cycle);
  }

  private static String buildMessage(String prefix, List<?> cycle) {
    StringBuilder message = new StringBuilder().append(prefix).append("\n");
    message.append("->").append(cycle.itemAt(0)).append("\n");
    for (int i = 1; i < cycle.count(); i++) {
      message.append("| ").append(cycle.itemAt(i)).append("\n");
    }
    message.append("--").append(cycle.itemAt(0));
    return message.toString();
  }

  public List<Object> cycle() {
    return cycle;
  }
}

File: src/tasks/model/impl/CyclicalDependencyException.kt
package tasks.model.impl

import java.lang.System.lineSeparator
import omnia.data.structure.List
import omnia.data.structure.immutable.ImmutableList

class CyclicalDependencyException internal constructor(msg: String, cycle: List<out Any>) : IllegalStateException(buildMessage(msg, ImmutableList.copyOf(cycle))) {
  val cycle: List<Any> = ImmutableList.copyOf(cycle)

  companion object {
    private const val serialVersionUID = -1193897049921304555L

    private fun buildMessage(prefix: String, cycle: List<*>): String {
      val message = StringBuilder().append(prefix).append(lineSeparator())
      message.append("->").append(cycle.itemAt(0)).append(lineSeparator())
      for (i in 1 until cycle.count()) {
        message.append("| ").append(cycle.itemAt(i)).append(lineSeparator())
      }
      message.append("--").append(cycle.itemAt(0))
      return message.toString()
    }
  }
}